docs/advanced.md

```diff
@@ -471,3 +471,160 @@ If you do need to make HTTPS connections to a local server, for example to test
 >>> r
 Response <200 OK>
 ```
+
+## Middleware
+
+Middleware is a general-purpose mechanism for extending the built-in functionality of a `Client`.
+
+### Using middleware
+
+Middleware generally comes in the form of classes. Middleware classes and their configuration parameters are meant to be passed as a list of `httpx.Middleware` instances to a client:
+
+```python
+from example.middleware import ExampleMiddleware
+import httpx
+
+middleware = [
+    httpx.Middleware(ExampleMiddleware, client_param="value", ...),
+]
+
+with httpx.Client(middleware=middleware) as client:
+    # This request will pass through ExampleMiddleware before
+    # reaching the core processing layers of the Client.
+    r = client.get("https://example.org")
+```
+
+### Writing middleware
+
+Middleware classes should accept a `get_response` parameter, that represents the inner middleware, as well as any keyword arguments for middleware-specific configuration options. They should implement the `.__call__()` method, that accepts a `request`, `timeout` configuration, and any per-request keyword arguments coming directly from keyword arguments passed to `client.get()`, `client.post()`, etc. The `__call__()` method should return a generator that returns a `Response` instance. To get a response from the inner middleware, use `response = yield from get_response(...)`.
+
+#### Basic example
+
+Here is a "do-nothing" middleware that sends the request unmodified, and returns the response unmodified:
+
+```python
+import httpx
+
+class PassThroughMiddleware:
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request, timeout, **kwargs):
+        return (yield from self.get_response(request, timeout, **kwargs))
+```
+
+#### Inspecting requests and responses
+
+Here is a middleware that prints out information about the request and the response, as well as any client or request options:
+
+```python
+import httpx
+
+class ExampleMiddleware:
+    def __init__(self, get_response, **options):
+        self.get_response = get_response
+        self.options = options
+        print(f"Client options: {self.options}")
+
+    def __call__(self, request, timeout, **kwargs):
+        print(f"Response options: {kwargs}")
+        print(f"Request: {request}")
+        response = yield from self.get_response(request, timeout, **kwargs)
+        print(f"Response: {response}")
+        return response
+
+middleware = [httpx.Middleware(ExampleMiddleware, client_option="example")]
+
+with httpx.Client(middleware=middleware) as client:
+    print("Sending request...")
+    r = client.get("https://example.org", request_option="example")
+    print("Got response")
+```
+
+Output:
+
+```console
+Client options: {'client_option': 'example'}
+Sending request...
+Request: <Request('GET', 'https://example.org')>
+Request options: {'request_option': 'example'}
```

Comment:
This might actually not be true, as `auth=None` and `allow_redirects=True` might end up in there.